metronome/checkpointing/app/src/io/iohk/metronome/checkpointing/app/CheckpointingComposition.scala

```diff
@@ -0,0 +1,572 @@
+
+  type RemoteNetworkMessage =
+    DuplexMessage[CheckpointingAgreement, CheckpointingMessage]
+
+  type NS = RocksDBStore.Namespace
+
+  type RNTS = NetworkTracers[Task, ECPublicKey, RemoteNetworkMessage]
+  type LNTS = NetworkTracers[Task, ECPublicKey, InterpreterMessage]
+  type CTS  = ConsensusTracers[Task, CheckpointingAgreement]
+  type STS  = SyncTracers[Task, CheckpointingAgreement]
+
+  /** Wire together the Checkpointing Service. */
+  def compose(
+      config: CheckpointingConfig
+  ): Resource[Task, Unit] = {
+
+    implicit val remoteNetworkTracers = makeRemoteNetworkTracers
+    implicit val localNetworkTracers  = makeLocalNetworkTracers
+    implicit val consesusTracers      = makeConsensusTracers
+    implicit val syncTracers          = makeSyncTracers
+    implicit val serviceTracer        = makeServiceTracer
+
+    for {
+      keyPair <- Resource.liftF(readPrivateKey(config)).map { privateKey =>
+        ECKeyPair(privateKey.underlying)
+      }
+
+      remoteConnectionProvider <- makeRemoteConnectionProvider(
+        config,
+        keyPair
+      )
+      remoteConnectionManager <- makeRemoteConnectionManager(
+        config,
+        remoteConnectionProvider
+      )
+      (hotstuffNetwork, applicationNetwork) <- makeNetworks(
+        remoteConnectionManager
+      )
+
+      localConnectionProvider <- makeLocalConnectionProvider(
+        config,
+        keyPair
+      )
+      localConnectionManager <- makeLocalConnectionManager(
+        config,
+        localConnectionProvider
+      )
+
+      db <- makeRocksDBStore(config)
+      implicit0(storeRunner: KVStoreRunner[Task, NS]) = makeKVStoreRunner(db)
+
+      blockStorage     <- makeBlockStorage(Block.genesis)
+      viewStateStorage <- makeViewStateStorage(Block.genesis)
+      ledgerStorage    <- makeLedgerStorage(config, Ledger.empty)
+
+      _ <- makeBlockPruner(config, blockStorage, viewStateStorage)
+
+      appService <- makeApplicationService(
+        config,
+        keyPair.pub,
+        applicationNetwork,
+        localConnectionManager,
+        blockStorage,
+        viewStateStorage,
+        ledgerStorage
+      )
+
+      _ <- makeHotstuffService(
+        config,
+        keyPair,
+        Block.genesis,
+        hotstuffNetwork,
+        appService,
+        blockStorage,
+        viewStateStorage
+      )
+
+    } yield ()
+  }
+
+  protected def makeRemoteNetworkTracers: RNTS =
+    CheckpointingRemoteNetworkTracers.networkHybridLogTracers
+
+  protected def makeLocalNetworkTracers: LNTS =
+    CheckpointingLocalNetworkTracers.networkHybridLogTracers
+
+  protected def makeConsensusTracers: CTS =
+    CheckpointingConsensusTracers.consensusHybridLogTracers
+
+  protected def makeSyncTracers: STS =
+    CheckpointingSyncTracers.syncHybridLogTracers
+
+  protected def makeServiceTracer =
+    CheckpointingServiceTracers.serviceEventHybridLogTracer
+
+  protected def makeConnectionProvider[M: Codec](
```

Comment:
Could we add explicit type annotations for the protected methods?

Reply:
Sure, I thought it would be self-describing and didn't want to add noise, but if you think it helps I'll add those.

Reply:
Done.